src/structstore_shared.hpp

```diff
@@ -42,62 +44,192 @@ class StructStoreShared {
-            shm_ptr = (SharedData*) mmap(nullptr, size, PROT_READ | PROT_WRITE,
-                                         MAP_SHARED, shm_fd, 0);
+
+            shm_ptr = (SharedData*) mmap(
+                    nullptr,
+                    size,
+                    PROT_READ | PROT_WRITE,
+                    MAP_SHARED,
+                    shm_fd,
+                    0);
+
             if (shm_ptr == MAP_FAILED) {
                 throw std::runtime_error("mmap'ing new memory failed");
             }
+
             // initialize data
+
             static_assert((sizeof(SharedData) % 8) == 0);
             new(shm_ptr) SharedData(size, bufsize, (char*) shm_ptr + sizeof(SharedData));
+
+            // marks the store as ready to be used
+            fchmod(shm_fd, 0660);
+
+        } else {
+            mmap_existing_shm();
         }
     }
 
+    void mmap_existing_shm () {
+
+        size_t size = sizeof(SharedData) + bufsize;
+        SharedData* original_ptr;
+
+        ssize_t result = read(shm_fd, &original_ptr, sizeof(SharedData*));
+        if (result != sizeof(SharedData*)) {
+            throw std::runtime_error("reading original pointer failed");
+        }
+
+        result = read(shm_fd, &size, sizeof(size_t));
+
+        if (result != sizeof(size_t)) {
+            throw std::runtime_error("reading original size failed");
+        }
+
+        if (size < sizeof(SharedData)) {
+            throw std::runtime_error("original size is invalid");
+        }
+
+        lseek(shm_fd, 0, SEEK_SET);
+        shm_ptr = (SharedData*) mmap(
+                original_ptr,
+                size,
+                PROT_READ | PROT_WRITE,
+                // ensures that the shared memory segment is mapped
+                // to the same region of memory for all processes
+                // the memory allocator relies on that
+                MAP_SHARED | MAP_FIXED_NOREPLACE,
+                shm_fd,
+                0);
+
+        if (shm_ptr == MAP_FAILED || shm_ptr != original_ptr) {
+            throw std::runtime_error("mmap'ing existing memory failed");
+        }
+
+        if (shm_ptr->original_ptr != original_ptr) {
+            throw std::runtime_error("inconsistency detected");
+        }
+
+        ++shm_ptr->usage_count;
+    }
+
+    bool verify () {
+
+        while (true) {
+
+            if (shm_fd == -1) {
+                shm_fd = shm_open(shm_path.c_str(), O_RDWR, 0600);
+
+                if (-1 == shm_fd) {
+                    return false;
+                }
+            }
+
+            if (shm_ptr == nullptr) {
+                struct stat shm_stat = {};
+                fstat(shm_fd, &shm_stat);
+
+                if (shm_stat.st_mode == 0100660) {
+                    mmap_existing_shm();
+                } else {
+                    return false;
+                }
+            }
+
+            if (!shm_ptr->ok.load()) {
+                munmap(shm_ptr, shm_ptr->size);
+                shm_ptr = nullptr;
```

Comment:
See comment above. Maybe try reopening the new shared mem first and replace the old only if successful? Maybe the `mmap_existing_shm()` could be a static method not modifying members and just returning a `shm_ptr`, this could simplify the usage here